server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java

```diff
@@ -119,6 +123,7 @@ public QueryShardContext(
         BitsetFilterCache bitsetFilterCache,
         TriFunction<MappedFieldType, String, Supplier<SearchLookup>, IndexFieldData<?>> indexFieldDataLookup,
         MapperService mapperService,
```

Comment:
We could. That require moving a couple other things to `MappingLookup` and it is the kind of thing I did in my first iteration but folks didn't like how "big" the abstraction got.

Reply:
I am really on the fence about this. I also would like to remove references to MapperService, yet I don't see a contained way of doing it, and I would not want to add all kinds of methods to MappingLookup. I wonder if it would be acceptable to consider this a (very important) follow-up, meaning that 7.11 will still have direct MapperService access while we are going to think about a long term solution for 7.x and master moving forward.

Reply:
I'd like to delay this simply because we aren't sure we want this. I think its reasonable but I understand the argument of wanting the `MapperService` reference for the "immutable stuff". It just that `MapperService` lets you do the mutable read too and it'd be nice not to have to worry about accidentally doing it.

If we had a separate object or interface for the immutable stuff I think it'd be really compelling to keep them separate.